File: 36-valid-sudoku/valid-sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        std::map<int,unordered_set<char>> r_map;
        std::map<int,unordered_set<char>> c_map;
        std::map<pair<int,int>, unordered_set<char>> rc_map;

        for (int i = 0; i< 9; i++){
            for(int j = 0; j < 9; j++){
                char elem = board[i][j];
                if (elem == '.') continue;
                else{
                    if(r_map[i].find(elem)!=r_map[i].end()) return false;
                    if(c_map[j].find(elem)!=c_map[i].end()) return false;
                    int r = i/3;
                    int c = j/3;
                    if(rc_map[{r,c}].find(elem)!=rc_map[{r,c}].end()) return false;
                    r_map[i].insert(elem);
                    c_map[j].insert(elem);
                    rc_map[{r,c}].insert(elem);
                }
            }
        }

        return true;
        
    }
};

File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        r_map: Defaultdict[int,Set[str]] = defaultdict(set)
        c_map: Defaultdict[int,Set[str]] = defaultdict(set)
        rc_map: Defaultdict[Tuple[int,int],Set[str]]  = defaultdict(set)

        for i in range(9):
            for j in range(9):
                elem = board[i][j]
                if elem=='.': continue
                else:
                    if elem in r_map[i]: return False
                    if elem in c_map[j]: return False
                    r = i//3
                    c = j//3
                    if elem in rc_map[(r,c)]: return False
                    r_map[i].add(elem)
                    c_map[j].add(elem)
                    rc_map[(r,c)].add(elem)
        
        return True

        